# Задача

Найти минимум функции $$f(x,y) = 6x^6 + 2x^4 y^2+10x^2 + 6xy+10y^2-6x +4 $$

# Решение

Найдём минимум функции программно с помощью градиентного спуска

In [88]:
def f(x):              # определяем функцию, минимум которой надо найти
    return 6*x[0]**6+2*x[0]**4 * x[1]**2 + 10*x[0]**2 + 6*x[0]*x[1] + 10 * x[1]**2 - 6*x[0] + 4

def gradf(x):    # градиент функции, минимум которой надо найти
    return [36*x[0]**5 + 8*x[0]**3*x[1]**2+20*x[0]+6*x[1]-6, 4*x[0]**4 * x[1] + 6*x[0]+20*x[1]]

def GD(lr, epsilon , f , gradf , x_0 ): # lr - шаг
    # epsilon - погрешность, на которые отличаются два последовательных приближения по евклидовой норме
    # f, gradf - функция и градиент функции
    #x_0 - начальное приближениe
    x_1 = []       
    x_2 = [0,0]
    x_1[:] = x_0[:]
    g = gradf(x_1)
    x_2[0] = x_1[0] - lr*g[0]  
    x_2[1] = x_1[1] - lr*g[1]
    err = ((x_2[0]-x_1[0])**2+(x_2[1]-x_1[1])**2)**(0.5)
    while err > epsilon :
        i+=1
        x_1[:] = x_2[:]
        g = gradf(x_1)
        x_2[0] = x_1[0] - lr*g[0]
        x_2[1] = x_1[1] - lr*g[1]
        err = ((x_2[0]-x_1[0])**2+(x_2[1]-x_1[1])**2)**(0.5)
    return x_2
        

In [102]:
x_0 = [0,0] # начальное приближение
lr = 0.01       # шаг
epsilon = 10**(-6) # погрешность

answ = GD(lr,epsilon, lambda x: f(x), lambda x: gradf(x) , x_0)

In [103]:
print(answ)

[0.32255398643866723, -0.09655428170480675]


# Ответ

Методом градиентного спуска получили приближённый минимум функции $$x = 0.32255 $$ $$y = -0.09655 $$